prov/opx/src/fi_opx_init.c

```diff
@@ -0,0 +1,481 @@
+	// Verify that pointers are exactly one cache-line long. */
+	OPX_COMPILE_TIME_ASSERT((sizeof(union fi_opx_hfi1_pio_state*) == 8),
+		"fi_opx_hfi1_pio_state pointer size error.");
+
+	OPX_COMPILE_TIME_ASSERT(sizeof(struct fi_opx_mr_atomic) == 24,
+							"Memory region Packet size error");
+
+	OPX_COMPILE_TIME_ASSERT(sizeof(struct fi_opx_mr_atomic) == sizeof(union fi_opx_mr_atomic_qw),
+							"Memory region Packet QW size error");
+
+	union fi_opx_hfi1_packet_payload *payload = NULL;
+	OPX_COMPILE_TIME_ASSERT(sizeof(*payload) == sizeof(payload->rendezvous.contiguous),
+							"Contiguous rendezvous payload size error");
+
+	OPX_COMPILE_TIME_ASSERT(sizeof(*payload) == sizeof(payload->rendezvous.noncontiguous),
+							"Non-contiguous rendezvous payload size error");
+}
+#pragma GCC diagnostic pop
+
+OPX_INI
+{
+	fi_opx_count = 1;
+	fi_opx_set_default_info(); // TODO: fold into fi_opx_set_defaults
+
+	if (fi_opx_alloc_default_domain_attr(&fi_opx_global.default_domain_attr)) {
+		return NULL;
+	}
+
+	if (fi_opx_alloc_default_ep_attr(&fi_opx_global.default_ep_attr)) {
+		return NULL;
+	}
+
+	if (fi_opx_alloc_default_tx_attr(&fi_opx_global.default_tx_attr)) {
+		return NULL;
+	}
+	if (fi_opx_alloc_default_rx_attr(&fi_opx_global.default_rx_attr)) {
+		return NULL;
+	}
+
+	fi_opx_global.prov = &fi_opx_provider;
+
+	fi_opx_init = 1;
+
+	fi_param_define(&fi_opx_provider, "uuid", FI_PARAM_STRING, "Globally unique ID for preventing OPX jobs from conflicting either in shared memory or over the OPX fabric. Defaults to \"%s\"",
+		OPX_DEFAULT_JOB_KEY_STR);
+	fi_param_define(&fi_opx_provider, "force_cpuaffinity", FI_PARAM_BOOL, "Causes the thread to bind itself to the cpu core it is running on. Defaults to \"No\"");
+	fi_param_define(&fi_opx_provider, "reliability_service_usec_max", FI_PARAM_INT, "The number of microseconds between pings for un-acknowledged packets. Defaults to 100 usec.");
+	fi_param_define(&fi_opx_provider, "reliability_service_pre_ack_rate", FI_PARAM_INT, "The number of packets to receive from a particular sender before preemptively acknowledging them without waiting for a ping. Valid values are powers of 2 in the range of 0-32,768, where 0 indicates no preemptive acking. Defaults to 0.");
+	fi_param_define(&fi_opx_provider, "selinux", FI_PARAM_BOOL, "Set to true if you're running a security-enhanced Linux. This enables updating the Jkey used based on system settings. Defaults to \"No\"");
+	fi_param_define(&fi_opx_provider, "hfi_select", FI_PARAM_STRING, "Overrides the normal algorithm used to choose which HFI a process will use. See the documentation for more information.");
+	fi_param_define(&fi_opx_provider, "delivery_completion_threshold", FI_PARAM_INT, "The minimum size message for doing full delivery completions. Smaller messages will provide injection completions.Value must be between %d and %d. Defaults to %d.", OPX_MIN_DCOMP_THRESHOLD, OPX_MAX_DCOMP_THRESHOLD, OPX_DEFAULT_DCOMP_THRESHOLD);
```

Comment:
These lines are too long. In general, try to keep each line under 80 columns unless there is good reason to make exception.